bitmap_amd64.go

```diff
@@ -1,69 +1,79 @@
-// Copyright (c) Roman Atachiants and contributors. All rights reserved.
-// Licensed under the MIT license. See LICENSE file in the project root for details.
-
-package bitmap
-
-import (
-	"github.com/klauspost/cpuid/v2"
-)
-
-var (
-	avx2 = cpuid.CPU.Supports(cpuid.AVX2)
-	popc = cpuid.CPU.Supports(cpuid.POPCNT)
-)
-
-// And computes the intersection between two bitmaps and stores the result in the current bitmap
-func (dst *Bitmap) And(b Bitmap) {
-	if dst.balance(b); len(*dst) >= len(b) {
-		switch avx2 {
-		case true:
-			x64and(*dst, b)
-		default:
-			and(dst, b)
-		}
-	}
-}
-
-// AndNot computes the difference between two bitmaps and stores the result in the current bitmap
-func (dst *Bitmap) AndNot(b Bitmap) {
-	if dst.balance(b); len(*dst) >= len(b) {
-		switch avx2 {
-		case true:
-			x64andn(*dst, b)
-		default:
-			andn(dst, b)
-		}
-	}
-}
-
-// Or computes the union between two bitmaps and stores the result in the current bitmap
-func (dst *Bitmap) Or(b Bitmap) {
-	if dst.balance(b); len(*dst) >= len(b) {
-		switch avx2 {
-		case true:
-			x64or(*dst, b)
-		default:
-			or(dst, b)
-		}
-	}
-}
-
-// Xor computes the symmetric difference between two bitmaps and stores the result in the current bitmap
-func (dst *Bitmap) Xor(b Bitmap) {
-	if dst.balance(b); len(*dst) >= len(b) {
-		switch avx2 {
-		case true:
-			x64xor(*dst, b)
-		default:
-			xor(dst, b)
-		}
-	}
-}
-
-// Count returns the number of elements in this bitmap
-func (dst Bitmap) Count() int {
-	if popc {
-		return int(x64count(dst))
-	}
-	return count(dst)
-}
+// Copyright (c) Roman Atachiants and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for details.
+
+package bitmap
+
+import (
+	"github.com/klauspost/cpuid/v2"
+)
+
+var (
+	avx2 = cpuid.CPU.Supports(cpuid.AVX2)
+	popc = cpuid.CPU.Supports(cpuid.POPCNT)
+)
+
+// And computes the intersection between two bitmaps and stores the result in the current bitmap
+func (dst *Bitmap) And(b Bitmap) {
+	var min int
+	if len(b) >= len(*dst) {
+		min = len(*dst)
+	} else {
+		min = len(b)
+		dst.shrink(min)
```

Comment:
I wanted to avoid shrink at all and use only reset bits, but didn't find the way to preserve capacity and decrease length. Do you have suggestion for that?

This shrink, i guess, is not good for 2 consecutive operations `dst.And(b)`, after that `dst.Or(c)` cause requires `grow` once again
```
*dst = make(Bitmap, blkAt+1, capacityFor(blkAt+1))
```

Reply:
I think it should be good, since `grow` can grow without copying if there is capacity and `shrink` implementation does not change capacity right now. It only resets bits to `0` and re-slices.

```go
func (dst *Bitmap) shrink(length int) {
	until := len(*dst)
	for i := length; i < until; i++ {
		(*dst)[i] = 0
	}

	// Trim without reallocating
	*dst = (*dst)[:length]
}
```

Reply:
Yeah, your are right, i have checked once again